contrib/memory/src/memory.nit

```diff
@@ -0,0 +1,994 @@
+		var ys = subh.to_f / img.height.to_f
+		scale = xs.min(ys)
+
+		# Compute the UI scale
+		xs = display.width.to_f / img.width.to_f
+		ys = display.height.to_f / img.height.to_f
+		ui_scale = xs.min(ys) / 4.0
+
+		var last = -1.0
+		var row = 0.0
+		var cpt = 0
+		for b in buttons do
+			b.place = cpt
+			var col = locs[cpt]
+			if col <= last then
+				row += 1.0
+			end
+			last = col
+
+			b.x = (col + 0.5) * slotw.to_f
+			b.y = (row + 0.5) * sloth.to_f
+			img = b.shape
+			img.scale = scale
+			b.w = (img.width.to_f * scale)
+			b.h = (img.height.to_f * scale)
+
+			cpt += 1
+		end
+
+		left.x = -150.0 * scale
+		left.y = (display.height / 2).to_f
+		right.x = display.width.to_f + 150.0 * scale
+		right.y = left.y
+
+		# Other UI elements
+
+		if not is_menu then
+			var reload = self.reload
+			drawing.reload.scale = ui_scale
+			reload.x = display.width.to_f - (drawing.reload.width.to_f / 2.0 * 1.2 ) * ui_scale
+			reload.y = drawing.reload.height.to_f / 2.0 * 1.2 * ui_scale
+			reload.w = drawing.reload.width.to_f * ui_scale
+			reload.h = drawing.reload.height.to_f * ui_scale
+		end
+	end
+
+	# The origin point of the cursor on the left
+	var left = new Pos(0.0, 0.0)
+
+	# The destination point of the cursor on the right
+	var right = new Pos(0.0, 0.0)
+
+	# The current cursor position
+	var cursor = new Pos(0.0, 0.0)
+
+	# The current cursor path
+	var path: nullable BPath = null
+
+	# The reload button
+	var reload: Button is noautoinit
+
+	# Safe point for a cursor on the i-th button of the level
+	fun path_pos(i: Int): Pos
+	do
+		if i < 0 then return left
+		if i >= level.length then return right
+		return level[i].to_pos
+	end
+
+	# A random point outside of the screen
+	fun far_away(display: Display): Pos
+	do
+		var a = (2.0*pi).rand
+		var w = display.width.to_f / 2.0
+		var h = display.height.to_f / 2.0
+		var x = w + a.cos * w * 1.8
+		var y = h + a.sin * h * 1.8
+		return new Pos(x, y)
+	end
+
+	# Create a BPath between two point with some nice handle values
+	fun new_path(from, to: Pos, ttl: Float): BPath
+	do
+		var a = atan2(to.y-from.y, to.x-from.x)
+		a += pi * (2.0.rand - 1.0)
+		var radius = 300.0 * scale
+		var fh = new Pos(from.x + a.cos*radius, from.y + a.sin*radius)
+		#var th = new Pos(to.x - a.cos*radius, to.y - a.sin*radius)
+		var path = new BPath(from, fh, to, to, ttl)
+		return path
+	end
+
+	# Initial placement of buttons
+	fun fly_in(display: Display)
+	do
+		for b in buttons do
+			var from = far_away(display)
+			var to = b.to_pos
+			var path = new_path(from, to, 5.0)
+			b.path = path
+			b.ttl = 5.0
+		end
+		ttl = 6.0
+	end
+
+	# Final leaving of buttons
+	fun fly_out(display: Display)
+	do
+		for b in buttons do
+			var from = b.to_pos
+			var to = far_away(display)
+			b.x = to.x
+			b.y = to.y
+			var path = new_path(from, to, 5.0)
+			b.path = path
+			b.ttl = 5.0
+			b.hard = false
+		end
+		ttl = 6.0
+	end
+
+	# Shuffle the buttons so that each on moves.
+	fun derangement
```

Comment:
`derangement` is the algorithmic concept: https://en.wikipedia.org/wiki/Derangement

shuffle is what is happening on the screen. But I can improve the doc